core/pta/apdu.c

```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (C) Foundries Ltd. 2021 - All Rights Reserved
+ * Author: Jorge Ramirez <jorge@foundries.io>
+ */
+#include <crypto/crypto_secure_element.h>
+#include <kernel/pseudo_ta.h>
+#include <pta_apdu.h>
+
+#define PTA_NAME "pta.apdu"
+
+static TEE_Result invoke_command(void *session_context __unused,
+				 uint32_t command_id, uint32_t pt,
+				 TEE_Param params[TEE_NUM_PARAMS])
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+						TEE_PARAM_TYPE_MEMREF_INPUT,
+						TEE_PARAM_TYPE_MEMREF_INPUT,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT);
+	TEE_Result ret = TEE_ERROR_NOT_IMPLEMENTED;
+	size_t len = 0;
+
+	FMSG("command entry point for pseudo-TA \"%s\"", PTA_NAME);
+
+	if (pt != exp_pt)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	switch (command_id) {
+	case PTA_CMD_TXRX_APDU_RAW_FRAME:
+		len = params[3].memref.size;
+		ret = crypto_secure_element_do_apdu(params[0].value.a,
+						    params[1].memref.buffer,
+						    params[1].memref.size,
+						    params[2].memref.buffer,
+						    params[2].memref.size,
+						    params[3].memref.buffer,
+						    &len);
+		if (!ret)
+			params[3].memref.size = len;
+		break;
+	default:
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+
+	return ret;
+}
+
+pseudo_ta_register(.uuid = PTA_APDU_UUID, .name = PTA_NAME,
+		   .flags = PTA_DEFAULT_FLAGS,
+		   .invoke_command_entry_point = invoke_command);
```

Comment:
APDU PTA can be freely called from REE clients without distinctions, or from TA clients? (no need for a open_session entry pointer?)

Reply:
not that I can think of...any TA client to REE client can send APDUs to the Secure Element at any time

Reply:
...and request or not key rotation, right?

Reply:
yes. In fact the the nxp agent linking to the libseteec might request rotation...and for sure the scp03 TA in uboot will need to be able to do it too.
see 
https://github.com/u-boot/u-boot/blob/master/cmd/scp03.c